RECIPE-84

PatsiChang opened the pull request:
Add Validations to Form Inputs
I am not sure how to pass the validation as array list from Text Input to Form Submit as each input fields requires individual and corresponding validations...
Link: https://smart-home-cloud.atlassian.net/browse/RECIPE-84

PatsiChang committed:
Add Validation to form input

ChakShingLo commented on new/components/basic/form/BasicTextInput.tsx:
```diff
@@ -1,26 +1,26 @@
-    label ?: string,
-    autoComplete ?: string
+    label?: string,
+    validationList?: ((input: any) => boolean)[],
```
avoid using a validation List. instead , the component should looks like
`<BasicTextInput required pattern="[a-zA-Z]" max="30"/>` which is a required field, accept letter only and 30 letters maximum.

The props interface should include all the supported validation.

ChakShingLo commented on new/lib/validations.tsx:
```diff
@@ -0,0 +1,48 @@
+
+//Check input is Number
+const validateInputIsNumber = (input: any) => {
+    return typeof input === "number";
```
input could be a string while the value is a number

ChakShingLo commented on new/components/basic/form/BasicForm.tsx:
```diff
@@ -1,25 +1,30 @@
+    onSubmitCallback?: (e: BaseSyntheticEvent, formData: FormData) => void,
 }
 
 export default function BasicForm({ onSubmitCallback, children, ...props }: FormProps) {
```
suggest to useContext and wrap the whole from with Provider. It will be a lot easier to capture the validation result as well as input value

PatsiChang committed:
Added validation

Only the required validation works for now, There is a bug for the submit description on Link 55- 58 BasicForm.tsx

PatsiChang committed:
Fixed bug

ChakShingLo commented on new/components/basic/form/ErrorCode.tsx:
```diff
@@ -0,0 +1,12 @@
+    return (
+        <h5>
+            {errorList.map((error, index) => <div key={index}>{`*${error}!`}</div>)}
+        </h5>
```
should use those Basic wrapper, or react-native `<Text>` if absent

PatsiChang replied:
Done

ChakShingLo commented on new/lib/validations.tsx:
```diff
@@ -0,0 +1,39 @@
+        if (input === undefined || input === null || input === "") {
+            return "is required";
+        }
+        return input?.length > 0;
```
1. why don't simply return true?

ChakShingLo commented on new/lib/validations.tsx:
```diff
@@ -0,0 +1,39 @@
+export const validationsMap: ValidationsMap = {
+    //Check required
+    "required": (input: any) => {
+        if (input === undefined || input === null || input === "") {
```
if the input contains space only, should it be considered as empty string as well?

ChakShingLo commented on new/lib/validations.tsx:
```diff
@@ -0,0 +1,39 @@
+    },
+    //Check input is String
+    "validateInputIsString": (input: any) => {
+        return (typeof input === "string") ? true : "Invalid Input";
```
this should validate the input value not the type, as user cannot alter the object type.

You may have to decide what input is considered as string in your application.

PatsiChang committed:
Fix PR bug

PatsiChang committed:
change validate one rule to many rules

PatsiChang committed:
Fix imports

PatsiChang committed:
migrate Error code after form submit

PatsiChang committed:
Fix Error code to prevent rerender when theres no error in login
